---
title: "GDSC Data Normalization"
output: 
  html_notebook: default
editor_options: 
  chunk_output_type: inline
---

```{r, include=F}
library(tidyverse)
library(magrittr)
library(disk.frame)
library(ggplot2)
library(here)
```

```{r, include=F}
knitr::opts_chunk$set(
  comment = '',
  fig.width = 6, 
  fig.height = 6
)

eval_r = TRUE
eval_bash = FALSE
setwd(here())

```

# Load Data:

**GDSC:**
* PanCancer ANOVA 
* PanCancer genetic features
* PanCancer IC50 results

**HUGO:**
* Gene synonyms

# ANOVA Results
Use the data from the direct page download. 


```{r}

gdsc_anova_raw = read.delim(
  here("large_files/gdsc/PANCANCER_ANOVA_Fri_Sep_24_18_40_17_2021.tsv"), 
  sep="\t",
  stringsAsFactors = F
  )

gdsc_genetic_features_raw = read.delim(
  here("large_files/gdsc/PANCANCER_Genetic_feature_Fri_Sep_24_18_41_14_2021.tsv"), 
  sep="\t",
  stringsAsFactors = F
  )

gdsc_ic50_raw = read.delim(
  here("large_files/gdsc/PANCANCER_IC_Fri_Sep_24_18_40_52_2021.tsv"), 
  sep="\t",
  stringsAsFactors = F
  )

gdsc_drug_list_raw = read.delim(
  here("large_files/gdsc/gdsc_drug_list.tsv"), 
  sep="\t",
  stringsAsFactors = F
  )

gdsc_cell_lines_raw = read.delim(here("large_files/gdsc/gdsc_cell_lines.tsv"), 
  sep="\t",
  stringsAsFactors = F
  )

gdsc_feature_list_raw = read.delim(
  here("large_files/gdsc/gdsc_feature_list.tsv"), 
  sep="\t",
  stringsAsFactors = F
  )

hugo_gene_synonyms_raw = read.delim(
  here("large_files/hugo/hugo.tsv"), 
  sep="\t",
  stringsAsFactors = F
  )

```

# Harmonize Column Names
* All lowercase (just easier typing in R)
* Replace "." or " " with "_"
* Add dataset type to "name". e.g. `name --> cell_name`
* Dataset size to "N" from "count" or "number"

## ANOVA
```{r}
gdsc_anova = gdsc_anova_raw
```

## Genetic Features
```{r}
gdsc_genetic_features = gdsc_genetic_features
```

## IC50
```{r}
gdsc_ic50 =
  gdsc_ic50_raw %>% 
  
  rename_with(function(x)
  {
    x = 
      x %>% 
      tolower() %>% 
      str_replace_all("\\.", "_")
  })
```

## Drug List
```{r}
gdsc_drug_list =
  gdsc_drug_list_raw %>%
  
  rename_with(function(x)
  {
    x = 
      x %>% 
      tolower() %>% 
      str_replace_all("\\.", "_")
  }) %>%
   
  rename(
    drug_name = name,
    drug_targets = targets,
    N = count,
    pubchem_id = pubchem
  )
```

## Cell List
```{r}
gdsc_cell_lines = 
  gdsc_cell_lines_raw %>% 
  
  rename(
    cell_line_name = Name,
    N = number
  )
```

## Feature List
```{r}
gdsc_feature_list = 
  gdsc_feature_list_raw %>% 
  
  rename_with(function(x)
  {
    x = 
      x %>% 
      tolower() %>% 
      str_replace_all("\\.", "_")
  }) %>% 
  
  rename(
    feature_type = type,
    genetic_feature = name,
    cna_driver_gene = known_driver_genes__cna_only_,
    N = count
  ) %>% 
  
  mutate(
    feature_type = recode(
      feature_type,
      "copy number alteration (loss)" = "copy number loss",
      "copy number alteration (gain)" = "copy number gain"
    )
  )

registerS3method("knit_print", "data.frame", print.me)
```




```{r}
intersect(colnames(gdsc_anova), colnames(gdsc_genetic_features))
```

```{r}
intersect(colnames(gdsc_anova), colnames(gdsc_ic50))
```

```{r}
intersect(colnames(gdsc_genetic_features), colnames(gdsc_ic50))
```

```{r}
list(
  drug_id = c("drug_id"),
  drug_name = c("Name"),
  synonyms = c("Synonyms"), 
  drug_targets = c("Targets"), 
  target_pathway = c("Target pathway"), 
  pubchem_id = c("PubCHEM"), 
  sample_size = c("Sample Size"),
  screening_site = c("Screening site"),
  feature_name = c("Feature", "Feature name"), 
  feature_type = c("Feature type"), 
  known_driver_genes = c("Known driver genes(CNA only)")
)
```


compounds
`Name`, `Synonyms`, `Targets`, `Target pathway`, `PubCHEM ID`, `Screening site`, `Datasets (cell line numbers)` 

drug_nae, synonyms, drug_targes, trug_target_pathway, pubchem_id, screening_site, gdsc_version, cell_line_sample_size



cancer features
 `Feature name`, `Feature type`, `Known driver genes(CNA only)`
 `feature_id`, `Name`, `Type`, `Known Driver Genes(cna only)`, `Count`
 
cancer_feature_id, cancer_feature_name, cancer_feature_type, known_driver_genes


cell lines
`Name`, `Cell Model Passports`, `COSMIC ID`, `TCGA Classification`, `Tissue`, `Tissue sub-type`, `Dataset (number of drugs)`

`COSMIC_ID`, `Name`, `COSMIC_ID`, `Tissue`, `Tissue sub-type`, `Count` 

cell_line_name, cell_model_passports_id, cosmic_id, tcga_classification, tissue, tissue_subtype, gdsc_version, drug_sample_size


anova
download page
`drug_name`, `drug_id`, `drug_target`, `target_pathway`, `feature_name`, `n_feature_pos`, `n_feature_neg`, `ic50_effect_size`, `log_ic50_mean_pos`, `log_ic50_mean_neg`, `log_max_conc_tested`, `log_max_conc_tested_2`, `feature_ic50_t_pval`, `feature_delta_mean_ic50`, `feature_pos_ic50_var`, `feature_neg_ic50_var`, `feature_pval`, `tissue_pval`, `msi_pval`, `fdr`, `tissue_type`, `screening set`

drug_name, drug_id, drug_target, target_pathway, feature_name, n_feature_pos, n_feature_neg, ic50_effect_size, log_ic50_mean_pos, log_ic50_mean_neg, log_max_conc_tested, log_max_conc_tested_2, feature_ic50_t_pval, feature_delta_mean_ic50, feature_pos_ic50_var, feature_neg_ic50_var, feature_pval, tissue_pval, msi_pval, fdr, tissue_type, screening set

ftp
`drug_name`, `drug_id`, `target`, `target_pathway`, `feature_name`, `n_feature_pos`, `n_feature_neg`, `ic50_effect_size`, `log_ic50_mean_pos`, `log_ic50_mean_neg`, `log_max_conc_tested`, `feature_ic50_t_pval`, `feature_delta_mean_ic50`, `feature_pos_ic50_var`, `feature_neg_ic50_var`, `feature_pval`, `tissue_pval`, `msi_pval`, `fdr`, `tissue_type`, `dataset_version`


drug_name, drug_id, drug_target, target_pathway, feature_name, n_feature_pos, n_feature_neg, ic50_effect_size, log_ic50_mean_pos, log_ic50_mean_neg, log_max_conc_tested, log_max_conc_tested_2, feature_ic50_t_pval, feature_delta_mean_ic50, feature_pos_ic50_var, feature_neg_ic50_var, feature_pval, tissue_pval, msi_pval, fdr, tissue_type, gdsc_version


drug data
`Drug name`, `Drug Id`, `Cell line name`, `Cosmic sample Id`, `TCGA classification`, `Tissue`, `Tissue sub-type`, `IC50`, `AUC`, `Max conc`, `RMSE`, `Z score`, `Dataset version`

`DATASET`, `NLME_RESULT_ID`, `NLME_CURVE_ID`, `COSMIC_ID`, `CELL_LINE_NAME`, `SANGER_MODEL_ID`, `TCGA_DESC`, `DRUG_ID`, `DRUG_NAME`, `PUTATIVE_TARGET`, `PATHWAY_NAME`, `COMPANY_ID`, `WEBRELEASE`, `MIN_CONC`, `MAX_CONC`, `LN_IC50`, `AUC`, `RMSE`, `Z_SCORE`

drug_name, drug_id, cell_line_name, cosmic_sample_id, tcga_classification, tissue, tissue_subtype, ic50, auc, max_concentration, rmse, z_score, gdsc_version, nlme_result_id, nlme_curve_id, cosmic_id, sanger_model_id, tcga_description, putative_target, pathway_name, company_id, webrelease, min_concentration, ln_ic50



cell_line_details
`Sample Name`, `COSMIC identifier`,	`Whole Exome Sequencing (WES)`,	`Copy Number Alterations (CNA)`, `Gene Expression`, `Methylation`,	`Drug Response`, `GDSC Tissue descriptor 1`, `GDSC Tissue descriptor 2`, `Cancer Type (matching TCGA label)`, `Microsatellite instability Status (MSI)`, `Screen Medium`, `Growth Properties`


cell_sample_name, cosmic_id, wes_analysis, copy_number_alteration, gene expression, methylation, drug_response, gdsc_tissue_descriptor_1, gdsc_tissue_descriptor_2, cancer_type, tcga_label, microsatellite_instability_status, screen_medium, growth_properties


cosmic tissue classificaiton
 `Line`, `COSMIC_ID`, `Site`, `Histology`
 
cell_line, cosmic_id, site_origin, histology
 

cell lines
`GDSC Name`, `COSMIC Name`, `COSMIC ID`
gdsc_name, cosmic_name, cosmic_id



 


