---
title: "Upload GDSC data to BQ"
author: "Ariel Balter"
date: "10/14/2021"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

# Knitr
```{r, include=F}
knitr::opts_chunk$set(
  warning=F,
  message=F,
  echo=F,
  error=F,
  number_section=T,
  include=F
)
```

# Libraries
```{r, include=F}
# Load Libraries
library(tidyverse)
library(magrittr)
library(ggplot2)
library(ggbeeswarm)
library(ggtext)
library(knitr)
library(printr)
# library(kableExtra)
library(flextable)
library(DT)
library(here)
library(DBI)
library(bigrquery)
library(jsonlite)
library(openxlsx)
```


# Custom table printing
```{r, include=F}
display_table = function(
  df,
  caption=NULL,
  search=list(regex = TRUE, caseInsensitive = TRUE)
)
{
  num_rows = nrow(df)
  df %>%
    datatable(
      extensions = c('ColReorder', 'Scroller'),
      options = list(
        colReorder=T,
        scroller=T,
        scrollY=300,
        sScrollX=T,
        searchHighlight = TRUE,
        search = search,

        pageLength = num_rows
      ),
      caption = htmltools::tags$caption(
        style = '
        caption-side: top;
        text-align: center;
        color:black;
        font-size:150% ;',
        caption),
      # caption = caption,
      rownames = T
    ) %>%
    formatStyle(columns = colnames(.), fontSize = '50%')
}

```


# Set Up BigQuery
```{r}
bq_deauth()
bq_auth(email="ariel.balter@providence.org")

# cancer_graph_project = "psjh-238522"
cancer_graph_project = "eacri-cancer-graph"
raw_dataset = "gdsc_raw"
```


# Create GDSC dataset
```{r}
# gdsc_dataset = bq_dataset(cancer_graph_project, dataset)
# 
# if (bq_dataset_exists(gdsc_dataset))
# {
#   bq_dataset_delete(gdsc_dataset, delete_contents = T)
# }
# 
# bq_dataset_create(gdsc_dataset)


gdsc_raw_conn = dbConnect(
  bigrquery::bigquery(),
  project = cancer_graph_project,
  dataset = raw_dataset
)
```


# Upload with Schema
## Upload Function
```{r}
arrayStringToStringArray = function(data)
{
  data = 
    data %>%
    replace_na(., "") %>% 
    str_split(., ",") %>% 
    lapply(str_trim)
  
  return(data)
}

arrayStringToJSON = function(data)
{
  data = 
    data %>%
    replace_na(., "") %>% 
    str_split(., ",") %>% 
    lapply(str_trim) %>% 
    unlist() %>% 
    toJSON()
  
  return(data)
}

df = data.frame(a = c("a,b , c", "d  , e"))

df %>% mutate(across(a, arrayStringToStringArray))
df %>% mutate(across(a, arrayStringToJSON))
df %>% rowwise() %>% mutate(a = arrayStringToStringArray(a))


arrayStringToStringArray(df$a)

type_map = list(
  INTEGER = "integer",
  FLOAT = "double",
  STRING = "character"
)


bqFieldsToColClasses = function(field_list)
{
  
  # print(fields)
  
  col_classes = lapply(field_list, function(field)
  {
    type_map[[field$type]];
  })
  
  names(col_classes) = lapply(field_list, function(field)
  {
    field$name
  })
  
  # print(col_classezs)
  
  return(col_classes)
}

#gsub("([^,])",'"\\1"', s) %>% sprintf("[%s]",.)

uploadToBQ = function(filename, schema, sep=",")
{
  
  repeated_fields = ""
    
  df = read.csv(
    file=here(filename),
    colClasses = bqFieldsToColClasses(schema),
    stringsAsFactors = F,
    sep=sep
  ) %>% 
  mutate(across(!repeated_fields, arrayStringToStringArray))
  
  table_bq = bq_table(
    project = gdsc_dataset$project, 
    dataset = gdsc_dataset$dataset, 
    table = basename(filename) %>% str_remove(".csv")
  )
  
  if (bq_table_exists(table_bq))
  {
    bq_table_delete(table_bq)
  }
  
  bq_table_upload(
    table_bq,
    quiet = FALSE,
    values = df,
    fields=schema
  )
}

```


# Metadata
## Drug Data

## Cell Line Data

## Genetic Features


# Results
## IC50

## Anova

## Genetic Features



## Drug Data Metadata
```{r}
drug_metadata_fields = list(
  bq_field(
    name="drug_name", 
    type="string", 
    description="TBD"
  ),
  bq_field(
    name="drug_id", 
    type="string", 
    description="From FAQ: \"The drug data presented at cancerrxgene.org is the result of a combined screening effort at the Massachusetts General Hospital (Boston, USA) and the Wellcome Sanger Institute (Cambridge, UK). Some drugs have been screened at both sites, resulting in two Drug IDs for the same drug. Drug ID values below 1000 indicate drugs that have been screened at MGH, while Drug IDs greater than 1000 were screened at Sanger.\n
    
Some associations will be overlapping between the pan-cancer and cancer-specific, whereas some will be unique to a specific cancer type. The results from these analyses can be queried separately using the drop-down menus on each results page.
Cancer-specific analyses were performed for 18 cancer types where n > 15 cell lines were available.\""
  ),
  bq_field(
    name="drug_targets", 
    type="string", 
    mode="repeated", 
    description="From FAQ: \"The target refers to the nominal therapeutic target(s) of a compound. In many, if not all, instances compounds have additional targets not listed here.\""
  ),
  bq_field(
    name="drug_target_pathways", 
    type="string", 
    mode="repeated", 
    description="From FAQ: \"\nThe target pathway/process has been manually curated based on our current understanding of cancer biology, therapeutic application and the biological processes in disease.\""
  ),
  bq_field(
    name="drug_name_synonyms", 
    type="string", 
    mode="repeated", 
    description="TBD"
  )
)

drug_metadata_fields = list(
  bq_field(name="drug_id", type="string", description="TBD"),
  bq_field(name="drug_name", type="string", description="TBD"),
  bq_field(name="drug_name_synonyms", type="string", mode="repeated", description="TBD"),
  bq_field(name="drug_targets", type="string", mode="repeated", description="TBD"),
  bq_field(name="drug_target_pathways", type="string", mode="repeated", description="TBD"),
  bq_field(name="pubchem_id", type="string", description="TBD"),
  bq_field(name="gdsc_release", type="string", description="TBD"),
  bq_field(name="sample_size", type="integer", description="TBD"),
  bq_field(name="screening_site", type="string", description="TBD")
)

drug_metadata_bq = tbl(gdsc_raw_conn, "gdsc_screened_drugs_metadata")

df = 
  drug_metadata_bq %>% 
  select(
    drug_id, 
    drug_name,
    drug_name_synonyms,
    drug_targets, 
    drug_target_pathways
  ) %>% 
  collect()


table_bq = bq_table(cancer_graph_project, graph_ready_dataset, "gdsc_drug_metadata")

if (bq_table_exists(table_bq))
{
  bq_table_delete(table_bq)
}

bq_table_upload(
  table_bq,
  quiet = FALSE,
  values = df,
  fields = drug_metadata_fields
)

```

## Genetic Features Metadata
```{r}
genetic_features_metadata_fields = list(
  bq_field(
    name="genetic_feature", 
    type="string", 
    description="Feature from unsupervised ML model. The ANOVA analysis currently correlates coding mutations and regions of recurrent copy number alteration with drug sensitivity data (IC50 values). These molecular alterations have been identified from the analysis of >11,000 patient tumours and subsequently mapped onto the cell lines for identifying molecular markers of drug sensitivity and resistance. "
  ),
  bq_field(
    name="genetic_feature_type", 
    type="string", 
    description="We identified regions of focal recurrent copy number alteration (amplification and homozygous deletion) from patient tumours, and these regions were then used to correlate with drug sensitivity observed in cancer cell lines. The pan-cancer set of segments was derived from the entire collection of cancer specific segments. CNA may contain many genes and in many cases the cancer gene(s) operative within the segment is unclear. For this reason, we report the association between a CNA and a specific compound, and any known cancer genes included within a CNA. Some segments do not contain known cancer genes. The complete list of genes included in each CNA is available on the relevant webpages, or in the download section of the website. "
  ),
  bq_field(
    name="known_driver_genes", 
    type="string", 
    mode="repeated", 
    description="Not sure. However, the list is unique to each genetic feature"
  )
)

df = 
  tbl(gdsc_raw_conn, "gdsc_genetic_features_metadata") %>% 
  select(
    genetic_feature = genetic_feature_name,
    genetic_feature_type,
    known_driver_genes
  ) %>% 
  collect()


table_bq = bq_table(cancer_graph_project, graph_ready_dataset, "gdsc_genetic_features_metadata")

if (bq_table_exists(table_bq))
{
  bq_table_delete(table_bq)
}

bq_table_upload(
  table_bq,
  quiet = FALSE,
  values = df,
  fields = genetic_features_metadata_fields
)
```

## Genetic Features Cell Line Metadata
```{r}
genetic_features_cell_line_metadata_fields = list(
  ### From genetic_features_metadata
  bq_field(
    name="genetic_feature",
    type="string",
    description="Feature from unsupervised ML model. We identified regions of focal recurrent copy number alteration (amplification and homozygous deletion) from patient tumours, and these regions were then used to correlate with drug sensitivity observed in cancer cell lines. The pan-cancer set of segments was derived from the entire collection of cancer specific segments. CNA may contain many genes and in many cases the cancer gene(s) operative within the segment is unclear. For this reason, we report the association between a CNA and a specific compound, and any known cancer genes included within a CNA. Some segments do not contain known cancer genes. The complete list of genes included in each CNA is available on the relevant webpages, or in the download section of the website. "
  ),
  bq_field(
    name="genetic_feature_type",
    type="string",
    description="What genetic features are included in the ANOVA analysis for associations with drug response? The ANOVA analysis currently correlates coding mutations and regions of recurrent copy number alteration with drug sensitivity data (IC50 values). These molecular alterations have been identified from the analysis of >11,000 patient tumours and subsequently mapped onto the cell lines for identifying molecular markers of drug sensitivity and resistance."
  ),
  bq_field(
    name="known_driver_genes",
    type="string",
    mode="repeated",
    description="Not sure. However, the list is unique to each genetic feature"
  ),

  ### From genetic_features
  bq_field(
    name="cell_line_name",
    type="string",
    description="TBD"
  ),
  bq_field(
    name="is_mutated",
    type="string",
    description="Most of the features are coded as not mutated (`is_mutated = 0`). I don't get this with feature_type. Or are only coding variants is_mutated=1, and not CNVs?"
  ),
  bq_field(
    name="recurrent_gain_loss",
    type="string",
    description="Not sure."
  ),
  bq_field(
    name="genes_in_segment",
    type="string",
    mode="repeated",
    description="We identified regions of focal recurrent copy number alteration (amplification and homozygous deletion) from patient tumours, and these regions were then used to correlate with drug sensitivity observed in cancer cell lines. The pan-cancer set of segments was derived from the entire collection of cancer specific segments. CNA may contain many genes and in many cases the cancer gene(s) operative within the segment is unclear. For this reason, we report the association between a CNA and a specific compound, and any known cancer genes included within a CNA. Some segments do not contain known cancer genes. The complete list of genes included in each CNA is available on the relevant webpages, or in the download section of the website. "
  )
)


genetic_features_metadata_bq = 
  tbl(gdsc_raw_conn, "gdsc_genetic_features_metadata") %>% 
  select(
    genetic_feature = genetic_feature_name,
    genetic_feature_type,
    known_driver_genes
  ) %>% 
  # mutate(known_driver_genes = array_to_string(known_driver_genes, "|")) %>% 
  # distinct() %>% 
  collect()

genetic_features_bq = 
  tbl(gdsc_raw_conn, "gdsc_genetic_features") %>% 
  select(
    cell_line_name,
    genetic_feature,
    genes_in_segment,
    is_mutated,
    recurrent_gain_loss
  ) %>% 
  # mutate(genes_in_segment = array_to_string(genes_in_segment, "|")) %>% 
  # distinct() %>% 
  collect()

df =
  full_join(
    genetic_features_metadata_bq,
    genetic_features_bq,
    by = "genetic_feature"
  ) %>% 
  # rowwise() %>% 
  mutate(
    known_driver_genes = ifelse(all(is.na(known_driver_genes)), c(""), known_driver_genes),
    genes_in_segment = ifelse(all(is.na(genes_in_segment)), c(""), genes_in_segment)
  )

  # mutate(across(genes_in_segment, toJSON)) %>% 
  # mutate(across(known_driver_genes, toJSON))


table_bq = bq_table(cancer_graph_project, graph_ready_dataset, "gdsc_genetic_features_cell_line_metadata")

if (bq_table_exists(table_bq))
{
  bq_table_delete(table_bq)
}

bq_table_upload(
  table_bq,
  quiet = FALSE,
  values = df,
  fields = genetic_features_cell_line_metadata_fields,
  compression = "GZIP"
)
```

## Cell Lines Metadata
```{r}
cell_lines_metadata_fields = list(
  bq_field(name="cell_line_name", type="string"),
  bq_field(name="tissue", type="string"),
  bq_field(name="tissue_subtype", type="string")
)

cell_lines_metadata_bq = 
  tbl(gdsc_raw_conn, "gdsc_cell_lines_metadata") %>% 
  select(
    cell_line_name,
    tissue,
    tissue_subtype
  )

df = cell_lines_metadata_bq %>% collect()

table_bq = bq_table(cancer_graph_project, graph_ready_dataset, "gdsc_cell_lines_metadata")

if (bq_table_exists(table_bq))
{
  bq_table_delete(table_bq)
}

bq_table_upload(
  table_bq,
  quiet = FALSE,
  values = df,
  fields = cell_lines_metadata_fields
)
```

## ANOVA
### Create Schema

```{r}
anova_fields = list(
  # bq_field(name="drug_name", type="string", description="TBD"),
  bq_field(name="drug_id", type="string", description="TBD"),
  # bq_field(name="drug_targets", type="string", mode="repeated", description="TBD"),
  # bq_field(name="target_pathway", type="string", description="TBD"),
  bq_field(name="genetic_feature", type="string", description="TBD"),
  # bq_field(name="n_feature_pos", type="integer", description="TBD"),
  # bq_field(name="n_feature_neg", type="integer", description="TBD"),
  bq_field(name="ic50_effect_size", type="float", description="TBD"),
  # bq_field(name="log_ic50_mean_pos", type="float", description="TBD"),
  # bq_field(name="log_ic50_mean_neg", type="float", description="TBD"),
  # bq_field(name="log_max_conc_tested", type="float", description="TBD"),
  # bq_field(name="log_max_conc_tested_2", type="float", description="TBD"),
  bq_field(name="feature_ic50_t_pval", type="float", description="The ANOVA calculates a p-value to determine the significance of each drug interaction. A Benjamini-Hochberg multiple testing correction was applied to the resulting p-values. A p-value threshold of <10-3 and a false discovery rate threshold equal to 25% were used to call significant associations across all the performed analyses.
  
In addition to the statistical significance, we advise you to investigate the effect size for each association and to prioritise genomic features associated with a large effect on drug sensitivity. This can be visualised using the scatter plot function. "),
    # bq_field(name="feature_delta_mean_ic50", type="float", description="TBD"),
  # bq_field(name="feature_pos_ic50_var", type="float", description="TBD"),
  # bq_field(name="feature_neg_ic50_var", type="float", description="TBD"),
  bq_field(name="feature_pval", type="float", description="TBD"),
  bq_field(name="tissue_pval", type="float", description="TBD"),
  bq_field(name="msi_pval", type="float", description="TBD"),
  bq_field(name="fdr", type="float", description="TBD")
  # bq_field(name="tissue_type", type="string", description="TBD"),
  # bq_field(name="gdsc_release", type="string", description="TBD")
)

```


### Upload
```{r}
# df = read.csv(
#   file=here("large_files/gdsc/gdsc_anova_results.csv"),
#   colClasses = bqFieldsToColClasses(anova_fields),
#   stringsAsFactors = F
#   ) %>% 
#   mutate(drug_targets = arrayStringToListArray(drug_targets))

# df = read.xlsx(
#   xlsxFile = 'large_files/gdsc/ANOVA_results_GDSC2_20Feb20.xlsx',
#   sheet = "PANCANCER_ANOVA"
# )
# df = 
#   df %>% 
#   rename(
#     drug_targets = target, 
#     gdsc_release = dataset_version) %>% 
#   mutate(trug_targets = arrayStringToListArray(drug_targets)) %>% 
#   mutate(drug_id = as.character(drug_id))

df = 
  tbl(gdsc_raw_conn, "gdsc_anova_results") %>% 
  select(
    drug_id,
    genetic_feature = feature_name,
    ic50_effect_size,
    feature_ic50_t_pval,
    feature_pval,
    tissue_pval,
    msi_pval,
    fdr
  ) %>% 
  collect()


table_bq = bq_table(cancer_graph_project, graph_ready_dataset, "gdsc_anova_results")

if (bq_table_exists(table_bq))
{
  bq_table_delete(table_bq)
}

bq_table_upload(
  table_bq,
  quiet = FALSE,
  values = df,
  fields=anova_fields
)
```


## Drug Data
```{r}
drug_ic50_data_fields = list(
  # bq_field(name="drug_name", type="string", description="TBD"),
  bq_field(name="drug_id", type="string", description="The drug data presented at cancerrxgene.org is the result of a combined screening effort at the Massachusetts General Hospital (Boston, USA) and the Wellcome Sanger Institute (Cambridge, UK). Some drugs have been screened at both sites, resulting in two Drug IDs for the same drug. Drug ID values below 1000 indicate drugs that have been screened at MGH, while Drug IDs greater than 1000 were screened at Sanger.
  
Some associations will be overlapping between the pan-cancer and cancer-specific, whereas some will be unique to a specific cancer type. The results from these analyses can be queried separately using the drop-down menus on each results page.

Cancer-specific analyses were performed for 18 cancer types where n > 15 cell lines were available."),
  bq_field(name="cell_line_name", type="string", description="TBD"),
  # bq_field(name="cosmic_sample_id", type="string", description="TBD"),
  # bq_field(name="tcga_classification", type="string", description="TBD"),
  # bq_field(name="tissue", type="string", description="TBD"),
  # bq_field(name="tissue_subtype", type="string", description="TBD"),
  bq_field(name="ic50", type="float", description="TBD"),
  bq_field(name="auc", type="float", description="TBD"),
  # bq_field(name="max_conc", type="float", description="TBD"),
  bq_field(name="rmse", type="float", description="The model error can be estimated by the root mean squared error (RMSE) which is indicative of how far the model approximation deviates from the measured observations. Where the RMSE is > 0.3 the fit is considered poor and the results have been filtered from further analysis. "),
  bq_field(name="z_score", type="float", description="TBD")
  # bq_field(name="gdsc_release", type="string", description="TBD")
)

df = read.csv(
  file=here("large_files/gdsc/gdsc_drug_ic50_data.csv"),
  colClasses = bqFieldsToColClasses(drug_ic50_data_fields),
  stringsAsFactors = F
  ) %>% 
  select(
    drug_id,
    cell_line_name,
    ic50,
    auc,
    rmse,
    z_score
  )

table_bq = bq_table(cancer_graph_project, graph_ready_dataset, "gdsc_drug_ic50_data")

if (bq_table_exists(table_bq))
{
  bq_table_delete(table_bq)
}

bq_table_upload(
  table_bq,
  quiet = FALSE,
  values = df,
  fields=drug_ic50_data_fields
)
```


## Cell Line Nodes
```{r}
genetic_features_cell_line_metadata_fields = list(
  bq_field(
    name="cell_line_name",
    type="string",
    description="TBD"
  ),
  bq_field(
    name="tissue_data",
    type="stringli",
    description="[{tissue_type: ?, tissue_subtype: ?, cancer_type: ?}]"
  ),
  bq_field(
    name="genetic_features",
    type="string",
    description="[{feature_name: ?, feature_type: ?, is_mutated: ?, recurrent_gain_loss: ?, known_driver_genes: ?, genes_in_segment: []}]"
  )
)


cell_lines_metadata = 
  tbl(gdsc_graph_conn, "gdsc_cell_lines_metadata") %>% 
  collect()

cell_lines_genetic_features = 
  tbl(gdsc_graph_conn, "gdsc_genetic_features_cell_line_metadata") %>% 
  collect()


df =
  full_join(
    genetic_features_metadata_bq,
    genetic_features_bq,
    by = "genetic_feature"
  ) %>% 
  # rowwise() %>% 
  mutate(
    known_driver_genes = ifelse(all(is.na(known_driver_genes)), c(""), known_driver_genes),
    genes_in_segment = ifelse(all(is.na(genes_in_segment)), c(""), genes_in_segment)
  )



  # mutate(across(genes_in_segment, toJSON)) %>% 
  # mutate(across(known_driver_genes, toJSON))


table_bq = bq_table(cancer_graph_project, graph_ready_dataset, "gdsc_genetic_features_cell_line_metadata")

if (bq_table_exists(table_bq))
{
  bq_table_delete(table_bq)
}

bq_table_upload(
  table_bq,
  quiet = FALSE,
  values = df,
  fields = genetic_features_cell_line_metadata_fields,
  compression = "GZIP"
)
```
